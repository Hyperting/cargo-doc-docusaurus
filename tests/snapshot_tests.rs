use cargo_doc_docusaurus::{converter, parser, ConversionOptions};
use std::path::Path;

#[test]
fn test_multifile_output_structure() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    let mut file_paths: Vec<_> = output.files.keys().collect();
    file_paths.sort();

    insta::assert_debug_snapshot!("multifile_structure", file_paths);
    assert_eq!(output.crate_name, "test_crate");
}

#[test]
fn test_index_file() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    let index_content = output.files.get("index.md").expect("index.md not found");
    insta::assert_snapshot!("index_md", index_content);
}

#[test]
fn test_lib_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    if let Some(lib_content) = output.files.get("index.md") {
        insta::assert_snapshot!("lib_module", lib_content);
    }
}

#[test]
fn test_types_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    if let Some(types_content) = output.files.get("types.md") {
        insta::assert_snapshot!("types_module", types_content);
    }
}

#[test]
fn test_functions_module() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    if let Some(functions_content) = output.files.get("functions.md") {
        insta::assert_snapshot!("functions_module", functions_content);
    }
}

#[test]
fn test_nested_modules() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    if let Some(nested_content) = output.files.get("nested.md") {
        insta::assert_snapshot!("nested_module", nested_content);
    }

    if let Some(inner_content) = output.files.get("nested/inner.md") {
        insta::assert_snapshot!("nested_inner_module", inner_content);
    }

    if let Some(deep_content) = output.files.get("nested/inner/deep.md") {
        insta::assert_snapshot!("nested_deep_module", deep_content);
    }
}

#[test]
fn test_struct_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    // Structs now use struct. prefix
    let unit_struct = output.files.get("struct.UnitStruct.md").expect("struct.UnitStruct.md not found");
    assert!(unit_struct.contains("title: \"Struct UnitStruct\""));
    assert!(unit_struct.contains("<RustCode"));

    let tuple_struct = output.files.get("struct.TupleStruct.md").expect("struct.TupleStruct.md not found");
    assert!(tuple_struct.contains("title: \"Struct TupleStruct\""));
    assert!(tuple_struct.contains("<RustCode"));

    let plain_struct = output.files.get("struct.PlainStruct.md").expect("struct.PlainStruct.md not found");
    assert!(plain_struct.contains("title: \"Struct PlainStruct\""));
    assert!(plain_struct.contains("### Fields"));
}

#[test]
fn test_generic_struct_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    let generic_struct = output.files.get("struct.GenericStruct.md").expect("struct.GenericStruct.md not found");
    assert!(generic_struct.contains("title: \"Struct GenericStruct\""));
    assert!(generic_struct.contains("### Generic Parameters"));
}

#[test]
fn test_enum_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    let simple_enum = output.files.get("enum.SimpleEnum.md").expect("enum.SimpleEnum.md not found");
    assert!(simple_enum.contains("title: \"Enum SimpleEnum\""));
    assert!(simple_enum.contains("### Variants"));

    let complex_enum = output.files.get("enum.ComplexEnum.md").expect("enum.ComplexEnum.md not found");
    assert!(complex_enum.contains("title: \"Enum ComplexEnum\""));
}

#[test]
fn test_function_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    // Functions now use fn. prefix and are in the functions module directory
    assert!(output.files.contains_key("functions/fn.add.md"));
    assert!(output.files.contains_key("functions/fn.multiply.md"));
}

#[test]
fn test_trait_formatting() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    let my_trait = output.files.get("trait.MyTrait.md").expect("trait.MyTrait.md not found");
    assert!(my_trait.contains("title: \"Trait MyTrait\""));
    assert!(my_trait.contains("```rust"));
}

#[test]
fn test_methods_and_impls() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    // Methods are now in individual struct files with struct. prefix
    let plain_struct = output.files.get("struct.PlainStruct.md").expect("struct.PlainStruct.md not found");
    assert!(plain_struct.contains("### Methods"));
}

#[test]
fn test_constants_and_type_aliases() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let output = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");

    let lib_content = output.files.get("index.md").expect("index.md not found");

    assert!(lib_content.contains("MAX_SIZE"));
    assert!(lib_content.contains("Result"));
}

#[test]
fn test_include_private() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");

    let output_public = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");
    let output_private = converter::convert_to_markdown_multifile(&crate_data, true, "", &[])
        .expect("Failed to convert to markdown");

    let public_count = output_public.files.len();
    let private_count = output_private.files.len();

    assert!(private_count >= public_count,
        "Private output should have at least as many files as public output");
}

#[test]
fn test_conversion_end_to_end() {
    let output_dir = std::env::temp_dir().join("cargo_doc_md_test");
    let _ = std::fs::remove_dir_all(&output_dir);

    let json_path = Path::new("tests/fixtures/test_crate.json");
    let options = ConversionOptions {
        input_path: json_path,
        output_dir: &output_dir,
        include_private: false,
        base_path: "",
        workspace_crates: &[],
    };

    cargo_doc_docusaurus::convert_json_file(&options).expect("Conversion failed");

    assert!(output_dir.join("test_crate").exists());
    assert!(output_dir.join("test_crate/index.md").exists());

    std::fs::remove_dir_all(&output_dir).ok();
}

#[test]
fn test_legacy_single_file_output() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    let markdown = converter::convert_to_markdown(&crate_data, false)
        .expect("Failed to convert to markdown");

    assert!(markdown.contains("# test_crate"));
    assert!(markdown.contains("## Table of Contents"));
    insta::assert_snapshot!("legacy_single_file", markdown);
}

#[test]
fn test_workspace_crates_internal_links() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    
    // First, let's check what external crates are referenced
    // We know from the JSON that there are references to external crates
    
    // Test without workspace_crates parameter
    let output_default = converter::convert_to_markdown_multifile(&crate_data, false, "", &[])
        .expect("Failed to convert to markdown");
    
    // Test with workspace_crates parameter  
    let workspace_crates = vec!["some_external_crate".to_string()];
    let output_with_workspace = converter::convert_to_markdown_multifile(
        &crate_data, 
        false, 
        "/docs/rust",
        &workspace_crates
    ).expect("Failed to convert to markdown");
    
    // Both should generate files successfully
    assert!(!output_default.files.is_empty(), "Default should generate files");
    assert!(!output_with_workspace.files.is_empty(), "With workspace_crates should generate files");
    
    // Verify base_path is used correctly in the second case
    let has_base_path_in_internal_links = output_with_workspace.files.values()
        .any(|content| {
            // Internal links to the current crate should use base_path
            content.contains("/docs/rust/test_crate/")
        });
    
    assert!(
        has_base_path_in_internal_links,
        "Internal links should use the provided base_path (/docs/rust)"
    );
    
    // Verify std library links still work (should always point to doc.rust-lang.org)
    let has_std_links = output_with_workspace.files.values()
        .any(|content| content.contains("doc.rust-lang.org"));
    
    assert!(
        has_std_links,
        "Std library links should always point to doc.rust-lang.org"
    );
}

#[test]
fn test_workspace_crates_with_hyphens() {
    let json_path = Path::new("tests/fixtures/test_crate.json");
    let crate_data = parser::load_rustdoc_json(json_path).expect("Failed to load JSON");
    
    // Test that workspace_crates works with hyphens (as in Cargo.toml)
    // even though crate names in JSON use underscores
    let workspace_crates_with_hyphens = vec!["test-crate".to_string()];
    let workspace_crates_with_underscores = vec!["test_crate".to_string()];
    
    let output_hyphens = converter::convert_to_markdown_multifile(
        &crate_data,
        false,
        "/docs",
        &workspace_crates_with_hyphens
    ).expect("Should work with hyphens");
    
    let output_underscores = converter::convert_to_markdown_multifile(
        &crate_data,
        false,
        "/docs",
        &workspace_crates_with_underscores
    ).expect("Should work with underscores");
    
    // Both should generate the same number of files
    assert_eq!(
        output_hyphens.files.len(),
        output_underscores.files.len(),
        "Both hyphen and underscore notation should work identically"
    );
}
