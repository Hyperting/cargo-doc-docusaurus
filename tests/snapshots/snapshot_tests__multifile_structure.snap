---
source: tests/snapshot_tests.rs
expression: file_paths
---
[
    "async_example/fn.async_with_args.md",
    "async_example/fn.boxed_future.md",
    "async_example/fn.generic_async.md",
    "async_example/fn.returns_future.md",
    "async_example/fn.simple_async.md",
    "async_example/index.md",
    "async_example/struct.AsyncCounter.md",
    "async_example/struct.AsyncStruct.md",
    "async_example/trait.AsyncIterator.md",
    "async_example/trait.AsyncTrait.md",
    "constant.MAX_SIZE.md",
    "constant.MIN_SIZE.md",
    "constant.VERSION.md",
    "enum.ComplexEnum.md",
    "enum.GenericEnum.md",
    "enum.SimpleEnum.md",
    "errors/enum.CustomError.md",
    "errors/fn.chain_errors.md",
    "errors/fn.fallible_operation.md",
    "errors/fn.operation_with_context.md",
    "errors/index.md",
    "errors/struct.ErrorContext.md",
    "errors/trait.IntoContext.md",
    "errors/type.Result.md",
    "fn.const_function.md",
    "fn.function_with_args.md",
    "fn.function_with_result.md",
    "fn.generic_function.md",
    "fn.multiple_bounds.md",
    "fn.simple_function.md",
    "fn.unsafe_function.md",
    "functions/fn.add.md",
    "functions/fn.async_function.md",
    "functions/fn.complex_generics.md",
    "functions/fn.const_function.md",
    "functions/fn.filter.md",
    "functions/fn.for_each.md",
    "functions/fn.function_with_very_long_signature.md",
    "functions/fn.higher_order_function.md",
    "functions/fn.map.md",
    "functions/fn.multiply.md",
    "functions/fn.process_mut_slice.md",
    "functions/fn.process_slice.md",
    "functions/fn.unsafe_function.md",
    "functions/index.md",
    "index.md",
    "lifetimes/enum.LifetimeEnum.md",
    "lifetimes/fn.higher_ranked_trait_bound.md",
    "lifetimes/fn.lifetime_function.md",
    "lifetimes/fn.multiple_lifetimes.md",
    "lifetimes/index.md",
    "lifetimes/struct.BorrowedData.md",
    "lifetimes/struct.DoubleBorrow.md",
    "lifetimes/struct.LifetimeStruct.md",
    "lifetimes/struct.LifetimeWithBound.md",
    "lifetimes/trait.LifetimeTrait.md",
    "nested/index.md",
    "nested/inner/deep/deeper/index.md",
    "nested/inner/deep/deeper/struct.DeeperStruct.md",
    "nested/inner/deep/fn.deep_function.md",
    "nested/inner/deep/index.md",
    "nested/inner/deep/struct.DeepStruct.md",
    "nested/inner/fn.inner_function.md",
    "nested/inner/index.md",
    "nested/inner/struct.InnerStruct.md",
    "nested/reexport_test/enum.GlobEnum.md",
    "nested/reexport_test/fn.glob_function.md",
    "nested/reexport_test/index.md",
    "nested/reexport_test/items/enum.GlobEnum.md",
    "nested/reexport_test/items/fn.glob_function.md",
    "nested/reexport_test/items/index.md",
    "nested/reexport_test/items/struct.GlobStruct.md",
    "nested/reexport_test/struct.GlobStruct.md",
    "nested/struct.OuterStruct.md",
    "patterns/index.md",
    "patterns/struct.Builder.md",
    "patterns/struct.Built.md",
    "patterns/struct.Closed.md",
    "patterns/struct.Handle.md",
    "patterns/struct.Newtype.md",
    "patterns/struct.Open.md",
    "patterns/struct.TypeState.md",
    "patterns/struct.Visitor.md",
    "struct.BoundedGeneric.md",
    "struct.Error.md",
    "struct.GenericStruct.md",
    "struct.PlainStruct.md",
    "struct.TupleStruct.md",
    "struct.UnitStruct.md",
    "trait.DisplayDebug.md",
    "trait.MyTrait.md",
    "traits/index.md",
    "traits/private/index.md",
    "traits/private/trait.SealedTrait.md",
    "traits/struct.AssociatedImpl.md",
    "traits/struct.SealedType.md",
    "traits/trait.Associated.md",
    "traits/trait.ComplexBounds.md",
    "traits/trait.Converter.md",
    "traits/trait.DefaultImpl.md",
    "traits/trait.Display.md",
    "traits/trait.ExtensionTrait.md",
    "traits/trait.FromIterator.md",
    "traits/trait.GenericTrait.md",
    "traits/trait.Iterator.md",
    "traits/trait.Sealed.md",
    "traits/trait.SuperTrait.md",
    "type.GenericResult.md",
    "type.Result.md",
    "types/constant.DEFAULT_CAPACITY.md",
    "types/constant.MAX_RETRIES.md",
    "types/enum.Status.md",
    "types/index.md",
    "types/struct.Container.md",
    "types/struct.Pair.md",
    "types/struct.RefStruct.md",
    "types/type.Map.md",
    "types/type.StringMap.md",
]
